### ISSUE TITLE
Develop new cron task for parsing extra data from scraped opinion PDFs

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#1141]
This goes hand in hand with freelawproject/juriscraper#283. It'll reorganize the code to move it out of CourtListener and into Juriscraper, and then this task is to use a cron job to call that code on any new items gathered during a given period of time (once/day?). 

I think the design should be pretty clear, one decision I leave open is whether this should be a generalized tool now or whether it should just solve the immediate problem. It could, for example, iterate over all our scrapers somehow and check them for PDF parsing code, or it could just leave if/else blocks for different courts...or do none of the above, leaving that for a later day.

@arderyp, I'd be curious if you have insight here. In the other issue we're finally making an API in Juriscraper to parse PDFs (or HTML if that's what we scrape), so that we can pull things out of the text of the opinions. In this one we're making a cron job to call that code. (We can't call it during the time of the scrape because the parsing of the text can be slow and is handled via async tasks.) Do you have an opinion about whether it's worth it at this stage to make the cron task generalized vs. just making it solve the immediate problem we're having with the Tax Court?

**Comment by johnhawkinson** [comment]
I think I have offered this before, but it's true and worth repeating: a cron job is the wrong solution for this sort of thing, and it is a _bad_ solution. It introduces unnecessary delay and turns what should be a triggered near-immediate (as much as possible) task into one with inordinate and (more) unpredictable delays.

It may well be that a cron job is the most immediate solution, and the best should not be the enemy of good enough, but we should realize it should never be a goal.

In this case, the right answer is a Queue of some sort, ideally processed immediately but which can stack up and be deferred when resources are limited. I don't remember if celery/asparagus/broccoli/some other vegetable makes any real attempt to solve this for you or not.

**Comment by arderyp** [comment]
I understand the view presented by @johnhawkinson, though I suppose I'm not as bullish on the point.  Installing/configuring/managing task runners is not a trivial task, so I think cron jobs work just fine for certain specific scenarios.  There's a time and a place for task runners, and maybe we're at that intersection now, or maybe not.  I personally have no clue what all is going on under the cron hood at FLP.  That being said, there's a whole world of possible scenarios out there where task runners are massive and major overkill.  I'll leave it at that.

Anyways, I read the other issue and this one and I'm still slightly confused, so bear with me.

As far as i understand, historically, we haven't been extracting text from opinions and saving them to the DB, partly because text mining PDFs can be tricky.  But, @flooie has written some code to consistently pull accurate text from PDFs, so now we want to add a new method to all scrapers that will return... what exactly?  An array of text blobs, one blob per each discovered opinion?  If so, is there some potential for reaching some limit there?  What if a scraper discovered hundreds of opinions, each consisting of lots and lots of pages of text.  The return array would be huge... are there theoretical or known limits we should be concerned about?

> Do you have an opinion about whether it's worth it at this stage to make the cron task generalized vs. just making it solve the immediate problem we're having with the Tax Court?

I don't entirely understand.  Are you asking if we should parse and store the text from Tax Court opinions only, or all opinions?  Does some specific processing need to be done on the Tax Court opinions (something particular need to be extracted)?  Do similar extractions need to happen for other court opinions?

Sorry if these are stupid questions, I don't really know what the source issue is regarding the Tax Court opinions and our need to parse/store their text.

**Comment by mlissner** [comment]
@arderyp, we always get the text, have for years. This one is about adding APIs for getting additional data *from that text*. For example, at the tax court, there's no way to get the danged docket number without parsing the text of the decision. So the question is, have we seen that kind of an issue elsewhere, and should we make a general solution that handles this or should we just handle the tax court issue for the moment?

(We do have a task runner, by the way. Celery.)

--------

@johnhawkinson We've got the Celery queue, so we could throw stuff there. The tricky part with this issue is that we're already using a queue to pull the text from the opinions. So if we just queue up a task to pull docket numbers (or whatever) out of the *text,* that queue will have to wait on the task in the regular queue to complete first. Celery *does* have a system where you can do what it calls a `chain`, in which you can set up several tasks to run one after another. We could probably do that, but it's tricky because we already use chains to do text extraction when OCR is needed, and...well, things are getting complicated. We felt like a slight delay was worth it here to avoid getting into that area of the code. This way it can live separately, not have dependencies on the other code, etc.

**Comment by mlissner** [comment]
Oh, one other thing, @johnhawkinson: I appreciate your continuing to hammer this point. Thank you.

**Comment by arderyp** [comment]
Ahhhh.  I knew as I was typing that it couldn't be right that we haven't been storing text historically.

Anyways, yes, we have seen scenarios before where the court doesn't display a docket number, or a case name in their opinion portal/listing.  In some of those cases we have to scrape supplemental html pages to find the info, and in others we simply don't track docket/name data because its only available in the PDF.  Unfortunately, I don't recall which scrapers exhibit these problems.  If I had to guess I'd say its under a dozen.  I think the issue of having to scrape supplemental pages is much more common than forgoing docket/title data altogether because its only available in the PDF. Here's an [example](https://github.com/freelawproject/juriscraper/blob/master/juriscraper/opinions/united_states/state/idahoctapp_u.py#L23-L35) of a scraper that has edge cases where only _some_ opinions are missing docket data (and I'm sure its listed in the opinion, but haven't checked).

As for implementing a more general generic solution for extracting certain data across the board, as opposed to addressing the specific Tax Court use case, I feel like I again must be misunderstanding something because wouldn't it just be insanely difficult if not impossible to write some common logic that will be able to successfully extract (for example) a docket number from any/every opinions we've ever come across.  Surely it can't be done, given the variation in text structuring, PDF formatting, and docket string structure.

Again, sorry if I'm missing something...

**Comment by arderyp** [comment]
@mlissner, also, I get your concern about chaining task, and creating child tasks that depend on parent tasks.  But wouldn't it be possible to create a task that simply monitors for new Tax Court opinions added to the database (with the extracted popinion text present and a docket value of null). This task could either handle the docket extraction itself or it could create a subtask for each opinion record, which would parse the text for said DB entry, find the docket number, and update the field in the DB.

**Comment by mlissner** [comment]
> because wouldn't it just be insanely difficult if not impossible to write some common logic that will be able to successfully extract (for example) a docket number from any/every opinions we've ever come across.

The idea we're tossing around in freelawproject/juriscraper#283 is:

1. A new hook in the base class that can be called from every opinion scraper. It takes the text of the PDF and returns an empty dictionary. It does nothing unless overridden like most of our hooks. 

2. In an individual scraper, you override the above method to return a dictionary with useful values. So you could have custom code that parses the opinion text as you expect it from that particular court, and then returns a dictionary with only the attributes you need parsed out. 

3. Then in this cron job, you can always call the method for a scraped PDF. Most of the time it'll do nothing, but if it's been overridden, it'll return a useful dictionary you can do things with.

> wouldn't it be possible to create a task that simply monitors for new Tax Court opinions added to the database

Isn't that polling and wouldn't that be very similar to a cron job anyway?

**Comment by arderyp** [comment]
So by "make the cron task generalized" you simply mean follows steps 1-3 above?  That is, implement a hook in the base class that returns an empty dict by default?  If so, then sure, there's no overhead, its dead simple to implements, and we can simply ignore it in most cases and opt-in where need be, which is already how a few of our hooks already work.  Sounds like a good approach to me that fits into the current structure nicely.

> Isn't that polling and wouldn't that be very similar to a cron job anyway?

Yes, hence my earlier statement that cronjobs are fine for certain tasks.

**Comment by flooie** [comment]
I think this makes sense.
> A new hook in the base class that can be called from every opinion scraper. It takes the text of the PDF and returns an empty dictionary. It does nothing unless overridden like most of our hooks.
